Implementation of cubic splines

rolandlo opened the pull request:
This commit implements cubic splines. At every knot you can adjust the tangent by click-dragging. The longer the drawn tangent, the stronger its influence. If all tangents are trivial (only click, no click-drag), it will be a linear spline as before.
The last set knot and the last set tangent can be modified using the keyboard (e, s, r, b for extend, shrink, rotate and back rotate).
If the spline is closed the tangent of the first knot will be used for the last knot as well, so the join will be smooth (if the tangent of the first knot is non-trivial).

rolandlo commented on src/model/SplineSegment.cpp:
```diff
@@ -0,0 +1,83 @@
+    secondControlPoint = q;
+}
+
+SplineSegment::SplineSegment(const Point& p, const Point& fp, const Point& sp, const Point& q) {
```
This suggested change gives me a segfault error upon execution. The error comes from invoking `SplineSegment firstPart = SplineSegment(firstKnot, b0, c0, d0);` in the `SplineSegment::subdivide`-method.

Technius replied:
The member initialization list should be equivalent to the assignments. Did you copy the above change verbatim? This constructor assigns `fp` and `sp` as well.

If the member initialization list syntax doesn't work, it's fine to just forget it.

rolandlo replied:
The constructor that does not work (although it compiles fine) is
```
SplineSegment::SplineSegment(const` Point& p, const Point& fp, const Point& sp, const Point& q): 
        firstKnot(p), firstControlPoint(fp), secondKnot(sp), secondControlPoint(q) {}

```
It does assign `fp` and `sp` as well. So I will leave it as it is.

Febbe replied:
I also tried it, and for me it works. There might be another problem. Do you tried to recompile the complete project?
```
SplineSegment::SplineSegment(const Point& p, const Point& fp, const Point& sp, const Point& q):
        firstKnot(p), secondKnot(q), firstControlPoint(fp), secondControlPoint(sp) {}
```

rolandlo replied:
Yes, I did. The program still crashes once I set the second knot (when the first tangent is non-trivial).

Febbe replied:
The thing is, Technius and my code sniped must not result in a segfault. Your version and our's also produce exactly the same binary output (with optimizations enabled). The only case it can crash there is, when one of the Point references are deleted. Since 2 of them are members of the SplineSegment, the SplineSegment may be deleted. And that would be the real Problem. I'll try to recompile it with sanitizers on and look for the real cause.

Febbe replied:
Ok I found the problem: It's a combination of using recursion, many growing vectors in a recursive function and the move into the return vector which may eliminate return value optimization. I'll list my change wish the following review comments.

rolandlo commented:
Thanks a lot for the feedback. I agree with everything except with the use of the member initialization list syntax in `SplineSegment::SplineSegment(const Point& p, const Point& fp, const Point& sp, const Point& q)`, see my comment above. Should I leave this constructor as is or change the code in another way?

rolandlo commented:
The follow-up commit contains some code-polishing (like using `->` consistently in function headings) as well. Are any more changes required?

Technius commented:
Just the one last code simplification above and then it's good to go!

Technius commented on src/control/tools/SplineHandler.cpp:
```diff
@@ -77,123 +94,113 @@ auto SplineHandler::onKeyEvent(GdkEventKey* event) -> bool {
+        }
+        case GDK_KEY_r:
+        case GDK_KEY_R: {  // r like "rotate"
+            double angle = (event->state == GDK_SHIFT_MASK) ? -5 : 5;
```
The event state is a bit field, so you need to use
```suggestion
            double angle = (event->state & GDK_SHIFT_MASK) ? -5 : 5;
```

Technius commented on src/control/tools/SplineHandler.cpp:
```diff
@@ -77,123 +94,113 @@ auto SplineHandler::onKeyEvent(GdkEventKey* event) -> bool {
+        }
+        case GDK_KEY_s:
+        case GDK_KEY_S: {  // s like "scale"
+            double factor = (event->state == GDK_SHIFT_MASK) ? 1 / 1.1 : 1.1;
```
Same here.
```suggestion
            double factor = (event->state & GDK_SHIFT_MASK) ? 1 / 1.1 : 1.1;
```

Febbe commented on src/model/SplineSegment.cpp:
```diff
@@ -0,0 +1,76 @@
+                          std::make_move_iterator(secondArray.end()));
+    }
+    return pointArray;
+}
```
Generally use early return as it increases readability, also it might be more efficient here.
Use "auto const&" for long constant return values or introduce a type alias and use it as "T const&".
Try to avoid to move into return value container, as it breaks possible return value optimisation. Instead assign it to a new local variable. 
The pop_back can be optimized out if the whole function only returns n-1 points.
The move_iterator is not necessary since Points are not movable.
Also vectors have a very bad scaling when resized often better use std::list or std::deque. 

Here an example:

```suggestion
auto SplineSegment::toPointSequence() const -> std::list<Point> {
    auto result = toPointSequenceImpl();
    result.emplace_back(secondKnot);
    return result;
}

auto SplineSegment::toPointSequenceImpl() const -> std::list<Point> {
    if (isFlatEnough()) {
        return {firstKnot};
    }
    auto const& childSegments = subdivide(0.5);
    auto left_points = childSegments.first.toPointSequenceImpl();
    left_points.splice(end(left_points), childSegments.second.toPointSequenceImpl());
    return left_points;
}
```

Febbe commented on src/model/SplineSegment.cpp:
```diff
@@ -0,0 +1,76 @@
+    return pointArray;
+}
+
+auto SplineSegment::subdivide(float t) -> std::pair<SplineSegment, SplineSegment> {
```
This function can be const:
```suggestion
auto SplineSegment::subdivide(float t) const -> std::pair<SplineSegment, SplineSegment> {
```

Febbe commented on src/model/SplineSegment.cpp:
```diff
@@ -0,0 +1,76 @@
+    return Point(p.x * (1 - t) + q.x * t, p.y * (1 - t) + q.y * t);
+}
+
+auto SplineSegment::evaluate(float t) -> Point {
```
This function is unused and can be deleted.
Also it can be const:
```suggestion
auto SplineSegment::evaluate(float t) const -> Point {
```

Febbe commented on src/model/SplineSegment.cpp:
```diff
@@ -0,0 +1,76 @@
+    return d0;
+}
+
+auto SplineSegment::isFlatEnough() -> bool {
```
```suggestion
auto SplineSegment::isFlatEnough() const -> bool {
```

Febbe commented on src/model/SplineSegment.h:
```diff
@@ -0,0 +1,114 @@
+class SplineSegment {
+public:
+    SplineSegment() = default;
+    ~SplineSegment() = default;
```
It doesn't really matter, but I prefer to remove the "custom defaulted destructor" as it makes no difference for the compiler, but many people not using our .clang-tidy file will get into a warning hell, because the "rule of 5" is not applied.

Febbe commented on src/model/SplineSegment.h:
```diff
@@ -0,0 +1,114 @@
+     * @brief draw the spline segment via the cairo_curve_to command
+     * @param cr the cairo context on which the spline segment ought to be drawn
+     */
+    void draw(cairo_t* cr);
```
```suggestion
    void draw(cairo_t* cr) const;
```

Febbe commented on src/model/SplineSegment.h:
```diff
@@ -0,0 +1,114 @@
+     * @param t the parameter between 0 and 1, which corresponds to the point where the spline is split
+     * @return A pair of two spline segements corresponding to the two parts of the spline
+     */
+    std::pair<SplineSegment, SplineSegment> subdivide(float t);
```
```suggestion
    std::pair<SplineSegment, SplineSegment> subdivide(float t) const;
```

Febbe commented on src/model/SplineSegment.h:
```diff
@@ -0,0 +1,114 @@
+     * @param t the parameter between 0 and 1, which corresponds to the point on the spline segment
+     * @return A point on the spline segment which corresponds to the parameter value t
+     */
+    Point evaluate(float t);
```
```suggestion
    Point evaluate(float t) const;
```

Febbe commented on src/model/SplineSegment.h:
```diff
@@ -0,0 +1,114 @@
+     * @brief checks if the spline segment is flat enough so that it can be drawn as a straight line
+     * @return true, if the spline segment is flat enough; false otherwise
+     */
+    bool isFlatEnough();
```
```suggestion
    bool isFlatEnough() const;
```

Febbe commented on src/model/SplineSegment.h:
```diff
@@ -0,0 +1,114 @@
+     * @brief Convert the spline segment to a vector of points.
+     * @return A point list which represents the spline segment without the start point.
+     */
+    std::vector<Point> toVectorOfPoints();
```
Here the declaration of my suggestion from above:
Also I did not understand, why the first point is missing.

```suggestion
    /**
     * @brief Convert the spline segment to a vector of points.
     * @return A point list which represents the spline segment.
     */
    std::list<Point> toPointSequence() const;

    /**
     * @brief Convert the spline segment to a vector of points.
     * @return A point list which represents the spline segment without the end point.
     */
    auto toPointSequenceImpl() const -> std::list<Point>;
```

rolandlo replied:
That "without the start point" comment was a mistake (I planned it differently earlier).

rolandlo commented:
Thanks, Febbe, for all your suggestions and explanations. I am sorry to say, however, that even with all those improvements the member initialization list gives me a segfault error upon execution. That's why I left the constructor with assignments.

rolandlo commented on src/control/tools/SplineHandler.cpp:
```diff
@@ -202,51 +215,111 @@ void SplineHandler::finalizeSpline() {
+    size_t indexLast = 0;
+    for (auto s: segments) {
+        for (auto p: s.toPointSequence()) {
+            if (validMotion(p, stroke->getPoint(indexLast))) {
```
I now added such a check in the flatness criterion, so distances between two subsequent points won't get too small.

rolandlo commented on src/control/tools/SplineHandler.cpp:
```diff
@@ -202,51 +215,111 @@ void SplineHandler::finalizeSpline() {
-        stroke->deletePoint(pointCount - 1);
+    // convert collection of segments to stroke
+    stroke->deletePointsFrom(0);
+    stroke->addPoint(knots.front());
```
It is no longer needed now that the check on the distance to the previous point has been removed.

rolandlo commented on src/model/SplineSegment.cpp:
```diff
@@ -0,0 +1,64 @@
+}
+
+auto SplineSegment::isFlatEnough() const -> bool {
+    double eps = 1.001;
```
In practice an "average" spline segments will now be converted to a list of roughly 70 points. I hope this is fine.

Febbe replied:
Yes, that's fine, the best would be to make this a preference, but until this is not the case, we should make a good compromise between usability, pleasant appearance, and performance / energy consumption. Otherwise we get either tickets for "high energy consumption" or a bad "look and feel".

rolandlo commented:
One last thing: I changed the key binding for moving the last drawn knot around. Instead of the arrow keys (which are used for page navigation) I use the x- and y-keys (in combination with SHIFT for reversing its effect) to move the knot in x- resp. y-direction. It was really troublesome to move knots up and  down in documents with more than one page with the arrow keys.
Is everything alright now?

Febbe commented on src/control/tools/SplineHandler.cpp:
```diff
@@ -202,51 +215,111 @@ void SplineHandler::finalizeSpline() {
+            if (validMotion(p, stroke->getPoint(indexLast))) {
+                stroke->addPoint(p);
+                indexLast++;
+            }
```
Done

Febbe commented:
Mostly everything is fine, that with the keys is indeed a bug in your code: Your Key Handler does not  return true if the key has been handled, this should be changed for all Actions. For the change to x and y I would consume @Technius. I prefer still "Up" and "Down".

There are some stylistic things like the use of `container.size()>0` which should be `!container.empty()`. Or the use of magic numbers (constants). Which should be constexpr variables with a descriptive name. But that's nitpicking on high niveau.
I don't expect you to change this.

Do you know how to [squash](https://www.internalpointers.com/post/squash-commits-into-one-git) your commit into one? You have to force push the squashed commit(s) to your branch.
For fixups or clang format I prefer to just use fixup(f) instead of squash(s).  
That would be the last step before merge:
1. Clean your history up (squash or fixup commits)
2. Rebase onto current master.

Technius commented:
> I use the x- and y-keys (in combination with SHIFT for reversing its effect) to move the knot in x- resp. y-direction. It was really troublesome to move knots up and down in documents with more than one page with the arrow keys.

Using <kbd>x</kbd> and <kbd>y</kbd> is really bad UX, but I know what the problem is. Review comment following soon...

Technius commented on src/control/tools/SplineHandler.cpp:
```diff
@@ -77,123 +94,115 @@ auto SplineHandler::onKeyEvent(GdkEventKey* event) -> bool {
     }
-    this->redrawable->rerenderRect(x - 1, y - 1, w + 2, h + 2);
+    this->redrawable->repaintRect(rect.x - 1, rect.y - 1, rect.width + 2, rect.height + 2);
     return false;
```
This always returns false, which means that the event will continue to be propagated even when it has been handled here. You will need to return true if you've done something with the keypress.

rolandlo replied:
I switched back to arrow keys and corrected the return statements.

rolandlo commented:
First time I squashed my commits into one, rebased on current master and force-pushed the squashed commit to my branch. But it seems it worked, didn't it?

rolandlo commented:
@Technius Anything left to be done, until it can be merged?

Febbe commented:
Sorry I think neither @technius nor me had time to review a last time. I currently have to work for example😅. But it should be perfectly mergeable. I also think, that additional changes should be done in a fixup PR when there are more. Because your work seems to be very clean, good and ready as is.

Technius commented:
LGTM, merging now. Thanks for the PR!
